internal/git/util.go

```diff
@@ -51,92 +52,129 @@ func GetLatestManifests(
 	gitSync *v1alpha1.GitSync,
 ) (string, []string, error) {
 	logger := logging.FromContext(ctx).With("GitSync name", gitSync.Name, "repo", gitSync.Spec.RepoUrl)
-	manifests := make([]string, 0)
 
 	// Fetch all remote branches
 	err := fetchUpdates(ctx, client, gitSync, r)
 	if err != nil {
-		return "", manifests, err
+		return "", nil, err
 	}
 
 	// The revision can be a branch, a tag, or a commit hash
 	hash, err := getLatestCommitHash(r, gitSync.Spec.TargetRevision)
 	if err != nil {
 		logger.Errorw("error resolving the revision", "revision", gitSync.Spec.TargetRevision, "err", err, "repo", gitSync.Spec.RepoUrl)
-		return "", manifests, err
+		return "", nil, err
 	}
 
 	localRepoPath := getLocalRepoPath(gitSync)
 
-	// repoPath will be the path of clone repository + the path while needs to be deployed.
+	// deployablePath will be the path of cloned repository + the path while needs to be deployed.
 	deployablePath := localRepoPath + "/" + gitSync.Spec.Path
-	// validate the deployablePath path for its existence.
+	// validate the existence of path.
 	if _, err := os.Stat(deployablePath); err != nil {
-		return "", manifests, fmt.Errorf("invalid deployable path, err: %v", err)
+		return "", nil, fmt.Errorf("invalid repo path, err: %v", err)
 	}
 
-	// initialize the kustomize with KUSTOMIZE_BINARY_PATH if defined.
-	k := kustomize.NewKustomizeApp(deployablePath, gitSync.Spec.RepoUrl, os.Getenv("KUSTOMIZE_BINARY_PATH"))
-
-	// initialize the helm with HELM_BINARY_PATH if defined.
-	h := helm.New(deployablePath, os.Getenv("HELM_BINARY_PATH"))
-
 	sourceType, err := gitSync.Spec.ExplicitType()
 	if err != nil {
-		return "", manifests, err
+		return "", nil, err
 	}
 
+	var targetObjs []*unstructured.Unstructured
 	switch sourceType {
 	case v1alpha1.SourceTypeKustomize:
-		generatedManifest, err := k.Build(nil)
-		if err != nil {
-			logger.Errorw("can not build kustomize yaml", "err", err)
-			return "", manifests, err
+		k := kustomize.NewKustomizeApp(localRepoPath, gitSync.Spec.Path, argoGit.NopCreds{}, gitSync.Spec.RepoUrl, os.Getenv("KUSTOMIZE_BINARY_PATH"))
+		targetObjs, _, err = k.Build(nil, nil, nil)
+	case v1alpha1.SourceTypeHelm:
+		targetObjs, err = helmTemplate(localRepoPath, gitSync)
+	case v1alpha1.SourceTypeRaw:
+		targetObjs, err = rawTemplate(r, gitSync, hash)
+	}
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to build manifests, err: %v", err)
+	}
+
+	manifests := make([]string, 0)
+	for _, obj := range targetObjs {
+		if obj == nil {
+			continue
 		}
-		manifestData, err := SplitYAMLToString([]byte(generatedManifest))
+		manifestStr, err := json.Marshal(obj.Object)
```

Comment:
If we need to marshal here, we can actually change the signature of the method to return `[]*unstructured.Unstructured` instead so that in `toUnstructuredAndApplyAnnotation` we don't need `Unmarshal` again. HDYT?